### ISSUE TITLE
1.0 Roadmap

### ISSUE THREAD

**Issue by GrahamCampbell** [php-build/php-build#276]
Some things we need to do before the v1.0.0 tag:

From @CHH:

> I think that we should think about things to clean up for 1.0, like removing the support for Pyrus (remember, it was introduced before composer was a thing).

I also think the broken snapshot versions need addressing too before we release, see #269.

**Comment by CHH** [comment]
Linking the issue about Pyrus removal: #30

**Comment by gerardroche** [comment]
Add a checklist of issues:

- [  ] [Delete the destination folder on failed instal](https://github.com/php-build/php-build/issues/125)

**Comment by rogeriopradoj** [comment]
Question about Versioning : what atom (X.Y.Z) will move with new definitions for new PHP versions?

**Comment by GrahamCampbell** [comment]
I think a minor version should be ok, unless there's any reason we want to do semver?

**Comment by GrahamCampbell** [comment]
I misread this as "remove" btw. :)

**Comment by gerardroche** [comment]
@rogeriopradoj I'm not sure I understand. Are you asking what versioning scheme php-build use?

**Comment by GrahamCampbell** [comment]
> Are you asking what versioning scheme php-build use?

We need to decide if we want to use semver. If we do, removing definitions needs to be major, adding them minor, and bug fixes patch.

**Comment by GrahamCampbell** [comment]
Though, technically changing the xdebug version on existing versions is a bc break, so needs a major version bump, so semver probably isn't for us.

**Comment by gerardroche** [comment]
I'd prefer semver because it's easier to see where I stand with compatability. Major bumps for removing definitions sounds good to me. Likewise, minor bumps for adding them is good.  xdebug version changes would need a major bump yes. Do xdebug version on existing versions need to be changed often? Maybe there's a better way to deal with that.

**Comment by CHH** [comment]
> Though, technically changing the xdebug version on existing versions is a bc break, so needs a major version bump, so semver probably isn't for us.

I think that's the reason why the folks at ruby-build went with ISO8601, e.g. `20150420` as version numbers.

The most important thing should be, that we are able to have a much faster release cadence than before, so there's a stable php-build release out when a new security fix for PHP is released.

**Comment by gerardroche** [comment]
> I think that's the reason why the folks at ruby-build went with ISO8601, e.g. 20150420 as version numbers.

One of the main values of semver is to indicate breaking changes that might require attention when updating. 

Example: A suggestion I've made in another issue is do away with arguments in favour of options i.e. use `php-build --definition <definition> --prefix <prefix>` instead of `php-build <definition> <prefix>`. Removing the argument style of invocation will require a major bump in semver. Semver `~1.0` -> `~2.0` indicates the bc change but ISO8601 `20150420` -> `20150510` would not.

### Version bump cases

* Minor bump
  * Adding definitions
* Major bump
  * Removing definitions. **Is removing definitions always a major bump?** Removing a valid definition that works is certainly a major bump. But what if the definition is broken and can't be fixed by php-build so it is being removed. I actually see this case as a patch level bump.  

#### xdebug version changes

If updating the xdebug version *fixes* something which is broken then it's a patch level fix.

If xdebug used semver then changing the minor or patch level xdebug version would only require a minor bump for php-build.

Command line options e.g. `--with-xdebug (master|latest|semver|default=latest)`  

> Though, the most important thing should be, that we are able to have a much faster release cadence than before, so there's a stable php-build release out when a new security fix for PHP is released.

:+1:

**Comment by GrahamCampbell** [comment]
@gerardroche I see no reason why we can't use minor releases t indicate breaking changes.

**Comment by GrahamCampbell** [comment]
> Is removing definitions always a major bump?

Yes, even if they go broken by a change we can't control. That's why I think semver would be a bad idea for is.

**Comment by gerardroche** [comment]
Yeah, thinking about I think by the very nature of php-build it probably can't be 100% semver compliant. 

I think we should still use a semver-ish version number.

Having thought about it, even adding definitions to me is not neccessarily a minor bump. Being able to build php releases is the feature, the fact that a versioned release definition is missing is a patch level fix. However, for example, adding snapshot build definitions is a new feature i.e. building snapshots, minor bump, but once that feature has been added, then adding a missing snapshot definition is only a patch.

If semver-ish version number is used, then it's mostly about deciding what level is bumped in what cases.

How about:

- patch 

  - adding missing definitions of an already implemented feature e.g. ability to build snapshots is a feature, minor bump, but once that feature has been implemented, then adding a missing snapshot definition is only a patch level fix
  - removing definitions that are broken outside the scope of php-build's control
  - bumping xdebug versions that are broken outside the scope of php-build's control

- minor

  - adding extensions, plugins, new options, etc... things within the scope and control of php-build 

- major 

  - anything that *is* within the scope and control of php-build and breaks backwards compatability

**Comment by GrahamCampbell** [comment]
> removing definitions that are broken outside the scope of php-build's control

I hate the idea of making any kind of bc break in a patch release. IMO, this should deffo be minor or major.

**Comment by rogeriopradoj** [comment]
Sometimes I think that the management of any definitions at all should be taken in consideration for bumping versions of `php-build` tool itself!

(Just being too much radical here, I know! However http://semver.org/ is that radical, isn't?!)

I'm taking a look at https://github.com/sstephenson/ruby-build for getting more familiarized with their version system.

Great thread, guys!

**Comment by fabiang** [comment]
Just my two cents: if you put the definitions into a separate repository, you don't run into trouble when updating either php-build or the definition files. :smile:

**Comment by rogeriopradoj** [comment]
Hey, @fabiang , thanks for your suggestion. It could be a path to follow.

Ping @php-build/collaborators @php-build/owners !
